src/Func.cpp

```diff
@@ -2569,36 +2569,42 @@ vector<Expr> FuncRef::args_with_implicit_vars(const vector<Expr> &e) const {
             << "Value " << (i+1) << " in definition of \"" << func.name() << "\" is undefined.\n";
     }
 
-    CountImplicitVars count(e);
+    CountVars count(e);
     for (size_t i = 0; i < a.size(); i++) {
         a[i].accept(&count);
     }
 
-    if (count.count > 0) {
+    CountVars rhs_count(e);
+    if (!func.has_pure_definition() && implicit_placeholder_pos != -1) {
+        user_assert(rhs_count.var_count > 0)
```

Comment:
I don't understand this part. If the left-hand-side uses implicit vars, why does it matter whether or not the right hand side uses non-implicit vars? Why track the total number of vars at all?

Reply:
Because we want to prevent implicit var in l.h.s and scaler function on the r.h.s. `e.g f(_) = 0`
This PR still allow _ on the l.h.s to expand to zero vars. That seem to be a common pattern (e.g in local_laplacian app)
```
downx(x, y, _) = (f(2*x-1, y, _) + 3.0f * (f(2*x, y, _) + f(2*x+1, y, _)) + f(2*x+2, y, _)) / 8.0f;
```
You can use this to downsample gray or rgb f.

Note:when we are here r.h.s implicit vars already expanded.